INGEST: Document Processor Conditional

original-brownbear opened the pull request:
Relates #33188

elasticmachine commented:
Pinging @elastic/es-core-infra

jakelandis commented on docs/reference/ingest/ingest-node.asciidoc:
```diff
@@ -721,12 +721,29 @@ All processors are defined in the following way within a pipeline definition:
+[source,js]
+--------------------------------------------------
+{
+  "bytes": {
```
Can we use the set processor for this example ?

original-brownbear replied:
Sure that seems less confusing :)

jakelandis commented on docs/reference/ingest/ingest-node.asciidoc:
```diff
@@ -721,12 +721,29 @@ All processors are defined in the following way within a pipeline definition:
+then the processor will be executed for the given document otherwise it will be skipped.
+The `if` field takes a map with the script settings used defined in <<script-processor, script-options>>
+and accesses a read only version of the document via the same `ctx` variable used by scripts in the
+<<script-processor>>.
```
Should we link to the painless doc ?

Do you think it would useful to also have additional examples (not necessarily the full the processor) of how to implement logstash's `~=` ? It seemed to be the most common operator to use for this type of check.

original-brownbear replied:
Hmm yea it could make sense to add more examples (I agree, it's way too hard to figure out how to do that in Painless from what we currently have in the docs).

WDYT @rjernst ?

rjernst commented on docs/reference/ingest/ingest-node.asciidoc:
```diff
@@ -721,12 +721,29 @@ All processors are defined in the following way within a pipeline definition:
 
 Each processor defines its own configuration parameters, but all processors have
-the ability to declare `tag` and `on_failure` fields. These fields are optional.
+the ability to declare `tag` ,`on_failure` and `if` fields. These fields are optional.
```
nit: space before comma should be after

rjernst commented on docs/reference/ingest/ingest-node.asciidoc:
```diff
@@ -721,12 +721,29 @@ All processors are defined in the following way within a pipeline definition:
 
+The `if` field must contain a script that returns a boolean value. If the script evaluates to `true`
+then the processor will be executed for the given document otherwise it will be skipped.
+The `if` field takes a map with the script settings used defined in <<script-processor, script-options>>
```
`if` doesn't take a map, right? I think you mean "object" (since we are talking about json), and that is optional (the example you have below here has `if` as the script source directly. You would only need an object if you wanted to pass params (unlikely for ingest I think) or use a different scripting language (not really possible since expressions don't support this, but theoretically a native script could be written as a custom script engine).

original-brownbear replied:
Right .. "object" wins :)

Hmm on that note and kinda off topic but: maybe we should/could enable expressions here the same way we do for bucket selector aggregation (interpret `1.0` as `true`) to become consistent there?

rjernst replied:
I think it would be difficult and not very useful, since expressions only have access to numerics. Expressions operate on doc values, but ingest does not have doc values, so it would require a lot of hacking.

original-brownbear replied:
ok makes sense :)

rjernst commented on docs/reference/ingest/ingest-node.asciidoc:
```diff
@@ -721,12 +721,29 @@ All processors are defined in the following way within a pipeline definition:
+  }
+}
+--------------------------------------------------
+// NOTCONSOLE
```
Why is this omitted from tests? Can we make an example that will work?

original-brownbear replied:
I figured just with the example config we don't have enough to go by for running anything, we'd need to actually index a document against a concrete pipeline here to make a test out of it right?
(that may be a little confusing given we only want an example of the `if` field here? Idk, you decide :))

rjernst replied:
IIRC we can have extra setup that is hidden from the generated documentation. I think having the examples always "work" is key to keeping the documentation up to date as apis change.

original-brownbear replied:
Ok :) Will look into that tomorrow morning :)

original-brownbear replied:
I can't add a test for a snipped like this it seems. I get this error:

```
Execution failed for task ':docs:buildRestTests'.
> reference/ingest/ingest-node.asciidoc[737:745](js)// CONSOLE: Didn't match (?:(?<comment>#.+)|(?:(?:(?<method>GET|PUT|POST|HEAD|OPTIONS|DELETE)\s+(?<pathAndQuery>[^\n]+)(?<body>(?:\n(?!GET|PUT|POST|HEAD|OPTIONS|DELETE|startyaml|#)[^\n]+)+)?)|(?:startyaml(?s)(?<yaml>.+?)(?-s)endyaml)))\n+: {
    "set": {
      "if": "ctx.bar == 'expectedValue'",
      "field": "foo",
      "value": "bar"
    }
  }
```

if I try anything here. => Seems I at least need a snippet that includes an actual request. That would break with the style of the following docs that also just show a quick outline of the configuration (without tests) for each processor?

original-brownbear commented:
@rjernst @jakelandis 

* Fixed space around comma
* Fixed wording "map" -> "object"
* Made the example use "set" instead of "bytes" proc

... but:

* Tests from the snippet in the form used here seem impossible with our test framework (unless I'm still misunderstanding something about it) ... we could use a different snippet though :)?

rjernst commented:
@nik9000 Can you provide any suggestions on making the snippet test discussed here work?

jakelandis commented:
@original-brownbear @rjernst - I think a standalone page (in addition to what is here), something like "Conditional Execution" would be beneficial. With specific mentions about the regex requirements, conditional dropping of events, and conditional pipelines. I can take the [TODO](https://github.com/elastic/elasticsearch/issues/33188) to create that content (and implement the tests there).  Hopefully that will unblock this PR. 

thoughts ?

original-brownbear commented:
@jakelandis I think that (standalone page) would make sense. Putting the example scripts and the how-to use the conditional processor could be pretty helpful for people trying to convert a LS config (or just their problem) into our code without having to jump between the Painless docs and pipeline docs :)

original-brownbear commented:
@rjernst alright thanks! Merged in master => will merge once green then.

jakelandis commented:
6.5 backport: https://github.com/elastic/elasticsearch/commit/fcad4e7889012b443bef62cf9513404a31a74481
